nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+
+To avoid leaking the identifier construction logic to the client, the handler would need a mapping of client identifiers
+to handler identifiers. Temporal can provide a strongly consistent store out of the box for this mapping.
+
+A single execution would need to map to multiple client identifiers to account for workflow starts and updates.
+
+Looking up an identifier in persistent storage added latency, and should be opted in to.
+
+### Handler generated identifiers
+
+With this approach, the handler returns a reference Id in response to a start call, which the caller should persist for
+use with the pre-defined RPCs as shown below:
+
+- `Start(operation, input, callbackURL, cancelBeforeStarted)` - handler responds inline or returns a reference Id
+  ([`*`](#cancel-before-started))
+- `Cancel(reference_id)` - cancel operation by handler provided reference Id
+- `GetResult(reference_id)` - get operation result by handler provided reference Id
+- `Describe(reference_id)` - get operation status by handler provided reference Id
+- `DeliverResult(token)` - deliver a result of an async operation using a token in the Start request's callback URL
+
+The handler _should_ either sign or qualify caller-provided reference Ids when dealing with untrusted sources and
+multi-tenant applications.
+
+In order to support stateless clients, the handler may provide an additional set of APIs to resolve client input to a
+reference Id. This may be something that is built into the protocol or implemented ad-hoc. To support this at the
+protocol level, we would need to add the following APIs:
+
+- `Resolve(operation, input) -> reference_id`
+
+-- OR --
+
+- `Cancel(operation, input)` - cancel operation by handler derived Id
+- `GetResult(operation, input)` - get operation result by handler derived Id
+- `Describe(operation, input)` - get operation status by handler derived Id
+
+NOTE: This looks a lot like the APIs for the "client generated Ids" approach, except the ID is structured instead of a
+single string.
+
+#### Cancel before started
+
+A caller may cancel an operation before it obtains a reference Id, e.g. if it is uncertain that the operation has been
+already started due to crashes and timeouts. Because the caller doesn't have a referencable Id or knowledge of which
+input fields are used to derive the Id, it would need to pass in the entire Start call input.
+
+This is especially important when we want to automate cancellations of workflow initiated operations, Temporal server
+cannot derive the Id from opaque input.
+
+### Comparing the approaches
+
+#### Handler Id qualification
+
+With handler generated Ids, to support stateless clients, the handler would need to be able to map an operation and
+input to a reference Id. It would also need to qualify a reference Id or securely issue the reference Ids when callers
+cannot be trusted. With caller generated Ids, the handler would only need to qualify an Id based on the caller context
```

Comment:
The handler ids are opaque for the client, as long as they are unique and unguessable, how this is done is handler implementation dependent. There are several ways to do that, but this should not be part of the spec...